### PR TITLE
Cleanup instrumentation backend

### PR DESCRIPTION
**Pull request by cd1m0**
The new addition of `forall` exposed 2 problems with the "design" of the original instrumentation code. Namely:

1. The design was centered on the assumption that transpiling can be split into 2 phases - a) flattening (breaking down a complex expression into a list of assignments and b) translation from Scribble to Solidity.
This assumption is now broken by `forall`-s - which require insertion of (potentially nested) loops with their own scopes, as well as potentially emitting assignments inside the loops as opposed to just at the top-level
2. Splitting re-writing into 2 parts - a) generating a list of rewrite steps (recipe) and b) executing (cooking) the recipe. 
This was originally a neat idea, however as the code grew, I found myself finding shortcuts away from building recipes due to the verbosity of the code. This half-way approach of sometimes using recipes, and sometimes not is confusing.

To fix those 2 problem this PR does the following:

1. Unify the old flattening (`flattenExpr`) and transpiling (`generateExprAst`) passes into a single transpiling (`transpile`) pass.
2. Remove all recipes and replace them with eager re-writes in-place.

### CODE REVIEW

**Comment by blitz-1306** [comment]
The other reason that **solc-typed-ast**'s `ASTNodeFactory` and ASTNode's insert/remove logic became a bit more reliable. So it is good decision to remove some obsolete code.

File: src/instrumenter/transpile.ts [hunk 1]
```diff
@@ -93,61 +98,57 @@ export function generateTypeAst(type: TypeNode, factory: ASTNodeFactory): TypeNa
...
+    if (BuiltinSymbols.has(expr.name)) {
+        return factory.makeIdentifier("<missing>", expr.name, -1);
     }
 
-    return DataLocation.Default;
-}
+    // Builtin symbols are the only identifiers with undefined `defSite`.
+    if (expr.defSite === undefined) {
+        throw new Error(
+            `Cannot generate AST for id ${expr.pp()} with no corrseponding definition.`
```
Comment (blitz-1306) [inline src/instrumenter/transpile.ts:115]: Typo in `corresponding`

**Review by blitz-1306 — DISMISSED** [review]
LGTM

File: src/instrumenter/transpile.ts (hunk 1 above)
Comment (cd1m0) [reply src/instrumenter/transpile.ts]: Fixed in f9494de

**Review by blitz-1306 — APPROVED** [review]

**Pull request merged**